Docs/feature_memory_protection.md

```diff
@@ -1,359 +1,210 @@
-stack, clear the PRESENT bit in the page table for the guard page of
-the Boot Strap Processor stack, and build the page tables. However, the stack switch handlers will still only be
-installed in DXE phase if CpuStackGuard is TRUE.
-
-### Overhead
-**O(1)** time and space.
-
-**Setting:**
-
-- If TRUE, UEFI Stack Guard will be enabled.
-
-## **Stack Cookies**
-[Stack Cookies](#stack-cookies) enable protection of the stack return pointer. The stack cookie value is specific to each loaded image and is generated at random on image load. Stack cookies are enabled at compile time, but if this
-setting is FALSE the interrupts generated by stack cookie check failures should be ignored.
-
-**Setting:**
-
-- If TRUE, stack cookie failures will cause a warm reset. If FALSE, stack cookie failure interrupts will be ignored.
-
-## **How to Set the Memory Protection Policy**
+## CPU Stack Guard
+
+CPU Stack Guard adds two additional pages to the stack base for each core. The first page is
+simply a guard page with the [EFI_MEMORY_RP](#efi_memory_rp) attribute. When a page fault
+occurs, the current stack address is invalid and so it is not possible to push the error
+code and architecture status onto the current stack. Because of this, there is a special
+"Exception Stack" or "Known Good Stack" which is the second page placed at the base
+of the stack. This page is reserved for use by the exception handler and ensures that a
+valid stack is always present when an exception occurs for error reporting.
+
+**A note on SMM:** An equivalent SMM stack guard feature is contained in
+[PiSmmCpuDxeSmm](https://github.com/tianocore/edk2/tree/master/UefiCpuPkg/PiSmmCpuDxeSmm)
+and is not dictated by this policy.
+
+**Note that the UEFI:** stack protection starts in DxeIpl, because the region is fixed,
+and requires
+[PcdDxeIplBuildPageTables](https://github.com/tianocore/edk2/blob/master/MdeModulePkg/MdeModulePkg.dec)
+to be TRUE. In Project Mu, we have hard-coded CpuStackGuard to be TRUE in PEI phase, so we
+always set up a switch stack and guard page in PEI. However, the stack switch handlers
+will still only be installed in DXE phase if CpuStackGuard is TRUE. If the stack guard is
+disabled in DXE, the paging attributes at the stack base will be removed during memory
+protection initialization.
+
+## Stack Cookies
+
+A stack cookie (also called stack canary) is an integer placed in memory just before the stack
+return pointer.Most buffer overflows overwrite memory from lower to higher memory addresses,
+so in order to overwrite the return pointer (and thus take control of the process) the canary
+value must also be overwritten. This value is checked to make sure it has not changed before a
+routine uses the return pointer on the stack.
+
+[Stack Cookies](#stack-cookies) enable protection of the stack return pointer. The stack cookie
+value is specific to each loaded image and is generated at random on image load. Stack cookies are

```

Comment:
```suggestion
value is specific to each loaded image and is generated at random on image load in DXE. Stack cookies are

```

Probably worth a StackCheckLib README pointer, once that PR is in.

Reply:
Agreed. I'll update this section once that PR is in

Reply:
Added a pointer to the StackCheckLib readme